tests/document.rs

```diff
@@ -1,4 +1,596 @@
+                wait_for_sync: None,
+                return_new: None,
+                return_old: None,
+                silent: None,
+                overwrite_mode: Some(DocumentOverwriteMode::Update),
+                overwrite: None,
+            }),
+        )
+        .await;
+
+    let result = update.unwrap();
+    assert_eq!(result.old.is_none(), true);
+    assert_eq!(result.new.is_none(), false);
+
+    let doc = result.new.unwrap();
+    assert_eq!(doc.document["no"], 4);
+
+    assert_eq!(result.header.is_none(), false);
+
+    let coll = database.drop_collection(collection_name).await;
+    assert_eq!(coll.is_err(), false);
+}
+
+#[maybe_async::test(
+    any(feature = "reqwest_blocking"),
+    async(any(feature = "reqwest_async"), tokio::test),
+    async(any(feature = "surf_async"), async_std::test)
+)]
+async fn test_get_read_document() {
+    test_setup();
+    let host = get_arangodb_host();
+    let user = get_normal_user();
+    let password = get_normal_password();
+
+    let collection_name = "test_collection_read_document";
+
+    let conn = Connection::establish_jwt(&host, &user, &password)
+        .await
+        .unwrap();
+    let mut database = conn.db("test_db").await.unwrap();
+
+    let coll = database.drop_collection(collection_name).await;
+    assert_eq!(coll.is_err(), true);
+
+    let coll = database.create_collection(collection_name).await;
+    assert_eq!(coll.is_err(), false);
+
+    let coll = database.collection(collection_name).await.unwrap();
+
+    let test_doc: Document<Value> = Document::new(json!({ "no":1 ,
+    "testDescription":"read a document"
+    }));
+
+    // First test is to read a simple document without options
+    let create = coll.create_document(test_doc, None).await;
+    assert_eq!(create.is_ok(), true, "succeed create a document");
+
+    let header = create.unwrap().header.unwrap();
+    let _key = header._key;
+    let _rev = header._rev;
+    let read = coll.read_document(_key.as_str()).await;
+
+    let result: Document<Value> = read.unwrap();
+
+    assert_eq!(result.document["no"], 1);
+    assert_eq!(result.document["testDescription"], "read a document");
+    // Test if we get the right doc when it does match
+    let read: Result<Document<Value>, ClientError> = coll
+        .read_document_with_options(
+            _key.as_str(),
+            Some(DocumentReadOptions::IfMatch(_rev.clone())),
+        )
+        .await;
+    assert_eq!(read.is_err(), false, "got the right document");
+    // Test if we get the 412 code response when there is no match
+    let read: Result<Document<Value>, ClientError> = coll
+        .read_document_with_options(
+            _key.as_str(),
+            Some(DocumentReadOptions::IfMatch("_dsdsds_d".to_string())),
+        )
+        .await;
+    // We should get a 412, for now for some reason the error is parsed as a document
+    // todo fix how the reponse/error is built
+    assert_eq!(read.is_err(), true, " we should get 412");
```

Comment:
Maybe we can print `read` so that we can easily find what exactly the server returned in CI.
```rust
assert_eq!(read.is_err(), true, " we should get 412, got: {:?}", read);
```

Reply:
Ok. Now I see the problem. The server return a typical json for error response.

So the point is to unify response that is a plain json object and that store all data in a `result` field.